landlab/components/network_sediment_transporter/sediment_pulser_each_parcel.py

```diff
@@ -0,0 +1,316 @@
+    [[1]
+    [3]
+    [3]
+    [5]
+    [5]
+    [5]
+    [2]]
+
+    """
+
+    _name = "SedimentPulserEachParcel"
+
+    _unit_agnostic = False
+
+    _info = {}  # works with the DataRecord
+
+    def __init__(self, grid, **kwgs):
+
+        SedimentPulserBase.__init__(self, grid, **kwgs)
+
+        """
+        instantiate SedimentPulserEachParcel
+
+        Parameters
+        ----------
+
+
+        grid : ModelGrid
+            landlab *ModelGrid* to place sediment parcels on.
+
+        **kwgs include:
+            parcels: landlab DataRecord
+                Tracks parcel location and attributes
+            D50: float, optional
+                median grain size [m]
+            D84_D50: float, optional
+                ratio of 84th percentile grain size to the median grain size
+            rho_sediment : float, optional
+                Sediment grain density [kg/m^3].
+            parcel_volume : float
+                parcel volume [m^3]
+            abrasion_rate: float
+                volumetric abrasion exponent [1/m]
+        """
+
+    def __call__(self, time, PulseDF=None):
+
+        """specify the location and attributes of each pulse of material added to
+        a Network Model Grid DataRecord
+
+        Parameters
+        ----------
+        time : integer or datetime64 value equal to nst.time
+            time that the pulse is triggered in the network sediment transporter
+        PulseDF : pandas dataframe
+            each row contains information on the deposition location and volume of
+            a single pulse of sediment. The pulse is divided into 'n' number of
+            parcels, where 'n' equals the np.ceil(pulse volume / parcel volume)
+            For details on the format of the DataFrame, see the docstring for
+            function _sediment_pulse_dataframe
+
+
+        Returns
+        -------
+            self._parcels
+                a DataRecord containing all information on each individual parcel
+        """
+
+        if (
+            PulseDF is None
+        ):  # if no PulseDF provided, raise error. Should at least provide an empty PulseDF
+            msg = "PulseDF was not specified"
+            raise ValueError(msg)
+
+        if (
+            PulseDF.empty is True
+        ):  # if empty, pulser stops, returns the existing parcels, call stops
+            return self._parcels
+            print("PulseDF is EMPTY")
+
+        variables, items = self._sediment_pulse_dataframe(
+            time,  # create variabels and and items needed to create the data record
+            PulseDF,
+        )
+
+        if self._parcels is None:  # if no parcels, create parcels
+            self._parcels = DataRecord(
+                self._grid,
+                items=items,
+                time=[time],
+                data_vars=variables,
+                dummy_elements={"link": [_OUT_OF_NETWORK]},
+            )
+            print("Parcels not provided, created a new DataRecord")

```

Comment:
Remove all the debugging statements.

Reply:
@keckje I made this a warning -- seems like a user should start with sediment on the river network.

Reply:
thanks